support-workers/src/main/scala/com/gu/zuora/subscriptionBuilders/GuardianWeeklySubscriptionBuilder.scala

```diff
@@ -1,65 +1,71 @@
 
-    val contractAcceptanceDate = Try(firstDeliveryDate.get) match {
-      case Success(value) => value
-      case Failure(e) => throw new BadRequestException(s"First delivery date was not provided. It is required for a Guardian Weekly subscription.", e)
-    }
+    val contractEffectiveDate = now()
+
+    val readerType = if (state.giftRecipient.isDefined) ReaderType.Gift else ReaderType.Direct
 
-    val recurringProductRatePlanId = validateRatePlan(weeklyRatePlan(guardianWeekly, environment, readerType), guardianWeekly.describe)
+    val recurringProductRatePlanId = validateRatePlan(weeklyRatePlan(state.product, environment, readerType), state.product.describe)
 
-    val promotionProductRatePlanId = if (isIntroductoryPromotion(guardianWeekly.billingPeriod, maybePromoCode)) {
-      weeklyIntroductoryRatePlan(guardianWeekly, environment).map(_.id).getOrElse(recurringProductRatePlanId)
+    val promotionProductRatePlanId = if (isIntroductoryPromotion(state.product.billingPeriod, state.promoCode)) {
+      weeklyIntroductoryRatePlan(state.product, environment).map(_.id).getOrElse(recurringProductRatePlanId)
     } else recurringProductRatePlanId
 
     val (initialTerm, autoRenew, initialTermPeriodType) = if (readerType == ReaderType.Gift)
-      (initialTermInDays(contractEffectiveDate, contractAcceptanceDate, guardianWeekly.billingPeriod.monthsInPeriod), false, Day)
+      (initialTermInDays(contractEffectiveDate, state.firstDeliveryDate, state.product.billingPeriod.monthsInPeriod), false, Day)
     else
       (12, true, Month)
 
-    val subscriptionData = buildProductSubscription(
-      requestId,
+    val subscriptionData = subscribeItemBuilder.buildProductSubscription(
       recurringProductRatePlanId,
-      contractAcceptanceDate = contractAcceptanceDate,
+      contractAcceptanceDate = state.firstDeliveryDate,
       contractEffectiveDate = contractEffectiveDate,
       readerType = readerType,
       autoRenew = autoRenew,
       initialTerm = initialTerm,
       initialTermPeriodType = initialTermPeriodType
     )
 
-    applyPromoCodeIfPresent(promotionService, maybePromoCode, country, promotionProductRatePlanId, subscriptionData)
+    applyPromoCodeIfPresent(
+      promotionService,
+      state.promoCode,
+      state.user.billingAddress.country,
+      promotionProductRatePlanId,
+      subscriptionData
+    ).map { subscriptionData =>
+      val soldToContact = state.giftRecipient match {
+        case None =>
+          SubscribeItemBuilder.buildContactDetails(Some(state.user.primaryEmailAddress), state.user.firstName, state.user.lastName, state.user.deliveryAddress.get, None)
+        case Some(gR) =>
+          SubscribeItemBuilder.buildContactDetails(gR.email, gR.firstName, gR.lastName, state.user.deliveryAddress.get, None)
+      }
+      subscribeItemBuilder.build(subscriptionData, state.salesforceContacts.recipient, Some(state.paymentMethod), Some(soldToContact))
```

Comment:
as with the others, creating the SubscribeItem doesn't call any external services, so I moved it here rather than in the handler.